pkg/lexer/lexer.go

```diff
@@ -0,0 +1,259 @@
+package lexer
+
+import (
+	"unicode"
+	"unicode/utf8"
+)
+
+type TokenType int
+
+// These are the main token types for shell
+const (
+	// Special tokens
+	ILLEGAL TokenType = iota
+	SEMICOLON
+	GREATER
+	LESS
+	GREATGREAT
+	LESSLESS
+	GREATAMPERSAND
+	LESSAMPERSAND
+	PIPE
+	AMPERSAND
+	IDENT
+	SINGLEQUOTE
+	DOUBLEQUOTE
+	BACKQUOTE
+	COMMENT
+	EOF
+)
+
+type Token struct {
+	Type TokenType
+	Val  string
+}
+
+// grammar
+// args -> IDENTS | empty
+// cmd and args -> IDENT args
+// pipe list -> PIPE cmd and args | cmd and args
+// io modifier -> GREAT IDENT | GREATGREAT IDENT | ...
+// io modifiers list -> io modifiers | empty
+// background optional -> ampersand | empty
+// cmd line -> pipe list io modifiers background optional NEWLINE | NEWLINE | ILLEGAL recovery.
+
+type Lexer struct {
+	input  string // input string
```

Comment:
Using bytes makes it difficult to work with uff-8 characters (I think). Unless you don't want to support utf-8 then its fine to use.

Reply:
Hmm, yeah we can work with that. But keep in mind that we won't be able to have utf-8 IDENTs later on (if we want to support that. Also, it results in us doing a lot of conversions inside the lexer to check whether the given byte matches a rune or vice versa

Reply:
Cant we just use [`utf8.DecodeRune`](https://pkg.go.dev/unicode/utf8#DecodeRune)?

Reply:
We can but it again takes more than one byte and checking each character (rune type) with byte string would require extreme care. I think it is fine to have it as string because in that case the conversion happens only once (when we call it) whereas if we were to do it here it would become cumbersome. What do you think?

Reply:
If thats the case we can use string.